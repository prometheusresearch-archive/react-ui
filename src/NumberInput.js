/**
 * @copyright 2016-present, Prometheus Research, LLC
 */

import * as React from 'react';
import {create} from './stylesheet';
import Input from './Input';

import {
  tryParseNumber,
  extractValueFromEvent
} from './FormUtils';

let stylesheet = create({
  Input: Input,
});

export default class NumberInput extends React.Component {

  static defaultProps = {
    stylesheet,
    value: '',
  };

  constructor(props) {
    super(props);
    this.state = {value: props.value};
  }

  render() {
    let {stylesheet: {Input}, ...props} = this.props;
    return (
      <Input
        {...props}
        value={this.state.value}
        onChange={this.onChange}
        />
    );
  }

  componentWillReceiveProps(nextProps) {
    let {value} = this.state;
    if (nextProps.value == null) {
      this.setState({value: ''});
    } else if (nextProps.value !== tryParseNumber(value)) {
      this.setState({value: String(nextProps.value)});
    }
  }

  onChange = event => {
    let value = extractValueFromEvent(event);
    this.setState({value: value || ''}, () => {
      if (value !== null) {
        value = tryParseNumber(value);
      }
      this.props.onChange(value);
    });
  }
}
