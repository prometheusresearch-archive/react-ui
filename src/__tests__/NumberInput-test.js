/**
 * @copyright 2016, Prometheus Research, LLC
 */

import Sinon from 'sinon';
import assert from 'power-assert';
import React from 'react';
import TestUtils from 'react-addons-test-utils';

import NumberInput from '../NumberInput';

describe('react-ui', function() {

  describe('<NumberInput />', function() {

    it('renders', function() {
      let root;
      let onChange = Sinon.spy();
      let renderer = TestUtils.createRenderer();
      renderer.render(
        <NumberInput onChange={onChange} />
      );
      root = renderer.getRenderOutput();
      assert(root.type === 'input');
      assert(root.props.value === '');
      root.props.onChange('12');
      assert(onChange.lastCall.args[0] === 12);
      root.props.onChange('12.');
      assert(onChange.lastCall.args[0] === 12);
      root.props.onChange('12.2');
      assert(onChange.lastCall.args[0] === 12.2);
      root.props.onChange('xx');
      assert(onChange.lastCall.args[0] === 'xx');
      root.props.onChange('');
      assert(onChange.lastCall.args[0] === '');
      renderer.render(
        <NumberInput onChange={onChange} value={undefined} />
      );
      root = renderer.getRenderOutput();
      assert(root.props.value === '');
    });
  });

});

