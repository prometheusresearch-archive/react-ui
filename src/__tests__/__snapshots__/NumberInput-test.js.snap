// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<input
  className="input-1197536361"
  disabled={undefined}
  onChange={[Function]}
  value=""
/>
`;

exports[`renders 2`] = `
<input
  className="input-1197536361"
  disabled={undefined}
  onChange={[Function]}
  value=""
/>
`;
